magma/types/decoupled.py

```diff
@@ -0,0 +1,167 @@
+import magma as m
+from magma.tuple import Product, ProductKind
+
+
+from magma.bit import Bit
+from magma.t import Type, In, Out
+from magma.tuple import Product, ProductKind
+
+
+class ReadyValidException(Exception):
+    pass
+
+
+class ReadyValidKind(ProductKind):
```

Comment:
Actually thought of another way to do this. That way you can ensure that the types are always valid. I also I see that a lot of the methods are only defined for one type or another (consumer or producer, but not both). This is a more natural way to handle that by creating two separate subclasses. Common/specialized methods can still go in the base class.

```python
import magma as m


def _process_info(info):
    if not isinstance(info, tuple):
        return info, None
    return info


class ReadyValidKind(m.ProductKind):
    def __getitem__(cls, info):
        T, direction = _process_info(info)
        fields = {"valid": m.Bit, "data": T, "ready": m.Bit}
        fields["_T_"] = T
        fields["_direction_"] = direction
        if direction is None or direction is m.Direction.Undirected:
            return type(f"ReadyValid[{T}]", (ReadyValid,), fields)
        if direction is m.Direction.In:
            fields["valid"] = m.In(fields["valid"])
            fields["data"] = m.In(fields["data"])
            fields["ready"] = m.Out(fields["ready"])
            return type(f"Producer(ReadyValid[{T}])", (ReadyValidProducer,), fields)
        if direction is m.Direction.Out:
            fields["valid"] = m.Out(fields["valid"])
            fields["data"] = m.Out(fields["data"])
            fields["ready"] = m.In(fields["ready"])
            return type(f"Producer(ReadyValid[{T}])", (ReadyValidConsumer,), fields)
        raise NotImplementedError(direction)

    def qualify(cls, direction):
        if cls._direction_ is not None:
            raise Exception()
        return cls[cls._T_, direction]


class ReadyValid(m.Product, metaclass=ReadyValidKind):
    def common_method(self):
        pass

    def specializable_method(self):
        pass


class ReadyValidProducer(ReadyValid):
    def producer_only_method(self):
        pass

    def specializable_method(self):
        pass


class ReadyValidConsumer(ReadyValid):
    def consumer_only_method(self):
        pass

    def specializable_method(self):
        pass


T = ReadyValid[m.Bits[10]]

proc = m.In(T)()
cons = m.Out(T)()

proc.common_method()
cons.common_method()
proc.specializable_method()
cons.specializable_method()
proc.consumer_only_method()

proc @= cons

def Producer(T): return m.In(T)
def Consumer(T): return m.Out(T)
```

The one downside is that `ReadyValid[...]` is not really a semantically valid type since it is undirected, where as in your construction even the base type is directed. Food for thought, not sure one is nec. better.

Reply:
One corner case is that `m.In(ReadyValid)` could be useful for monitoring signals

Reply:
I think separating the logic into subclasses would be fine if we choose to restrict the interface to Producer/Consumer (since these functions could return the corresponding subclasses), for the above corner case we could also add Monitor(RV[T]), dunno if there's ever a use case of a Driver(RV[T]) (where everything is outputs).  

Another thought is to consider using Producer/Consumer as the type itself then, since we wouldn't have a base RV[T] (as you noted it wouldn't really be valid/useful), so instead of we could just have `Producer[T]` and `Consumer[T]` which would generate the appropriate ready/valid interface. I think the one reason we wouldn't want to do this is because looking at the Chisel code, there's variations on ReadyValid, such as Irrevocable[T] (once valid is high, the producer cannot change the data until it has been consumed).  So having Producer[ReadyValid[T]) and Producer[Irrevocable[T]] would be useful for distinguishing the variations on the interface.

Perhaps we can view ReadyValid[T] as an abstract type or protocol (based on the interface ports), since it might be useful for structural equality or subclass relationships (check it's an RV[T] without caring about the port polarity or even subclass relationship).  The ReadyValidConsumer/Producer[T] classes would then be concrete implementations (with a similar pattern for irrevocable and other derivatives)

Reply:
I do like the idea of `ReadyValid` being a "structural" (no semantics) base type, and `Producer`, `Consumer`, `Irrecovable`, etc. being subtypes with semantics (i.e. functions like enqueue etc).

I think the ideal usage would look like:

```python
ReadyValid[T]            # undirected tuple, no member functions                                                                                       
In(ReadyValid[T])        # *all* inputs, analogous Out(.)                                                                                              
Monitor(ReadyValid[T])   # alias for In(.)                                                                                                             
Consumer(ReadyValid[T])  # properly directed (i.e. ready=In, valid=Out)                                                                                
                         # semantic member functions                                                                                                   
Producer(ReadyValid[T])  # properly directed (i.e. ready=Out, valid=In)                                                                                
                         # semantic member functions                                                                                                   
Consumer(ReadyValid[T]).flip()  # alias for Producer(.)
```

Imagine there's many ways to do this. Not sure what the best software pattern for this is.

Reply:
Another related question is if we disallow explicitly wiring to fields of ready valid's. E.g.

```python
x = Consumer(ReadyValid[T])()
x.enqueue(...)
x.data @= ...
```

I think it's hard (to impossible) to disallow this...but I guess we'd get an error/warning from downstream wire call anyway.

Reply:
I think exposing the fields is okay, perhaps the user wants to define their own helper functions that are more complex than enq/deq and this would allow them to wire up the drivers directly if needed.  I think we can document the interface methods in terms of how they perform wiring, and say that the user can wire these ports directly but are responsible for making sure it works if they combine this with the methods

Reply:
When we add support for conditional assignment (ala chisel when), then the deq/enq logic for when becomes quite simple to describe in terms of conditional assignment (that's where most of the complexity currently comes from)